migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/EntityVersionMapMongo.scala

```diff
@@ -0,0 +1,116 @@
+
+import scala.reflect.ClassTag
+
+final class EntityVersionMapMongo(db: MongoDatabase) extends EntityVersionMap {
+  private val log: Logger = LogManager.getLogger(this.getClass)
+
+  private val codecRegistry: CodecRegistry = fromRegistries(fromProviders(classOf[EntityVersionMapping]),
+    DEFAULT_CODEC_REGISTRY)
+
+  import ScalaMongoImplicits._
+
+  private val entityMapCollectionName = "entity_version_map"
+
+  ensureCollectionExists()
+
+  /**
+    * An implementation if this abstract class should redefine this method. Clients of this class should use add()
+    * as it does additional checks.
+    *
+    * @param collectionName The name of the collection that contains the entity
+    * @param entityName     An entity name
+    * @param oldVersion     An version of the entity in the old version of the database
+    * @param newVersion     An version of the entity in the new version of the database
+    */
+  override protected def addEntry(collectionName: String,
+                                  entityName: String,
+                                  oldVersion: Int,
+                                  newVersion: Int): Unit = {
+    getTypedCollection[EntityVersionMapping]
+      .insertOne(EntityVersionMapping(collectionName, entityName, oldVersion, newVersion))
+      .execute()
+  }
+
+  /**
+    * Gets a 'name - version' mapping.
+    *
+    * @param collectionName The name of the collection that containg the entity
+    * @param entityName     An entity name
+    * @param oldVersion     An version of the entity in the old version of the database
+    * @return An version of the entity in the new version of the database, None if the entity is not found
+    *         in the mapping
+    */
+  @throws[IllegalStateException]
+  override def get(collectionName: String,
+                   entityName: String,
+                   oldVersion: Int): Option[Int] = {
+    val mappings = getTypedCollection[EntityVersionMapping]
+      .find(getFilter(collectionName, entityName, oldVersion))
+      .execute()
+    if (mappings.lengthCompare(1) > 0) {
+      throw new IllegalStateException("Duplicate entries in entity version mappings found. " +
+        s"Values: $mappings.")
+    } else {
+      mappings.headOption.map(record => record.newVersion)
+    }
+  }
+
+  /**
+    * Returns a collection that is serialized as a case class.
+    */
+  private def getTypedCollection[T](implicit ct: ClassTag[T]): MongoCollection[T] = {
+    db.getCollection[T](entityMapCollectionName)
+      .withCodecRegistry(codecRegistry)
+  }
+
+  /**
+    * Returns a filter for the MongoDB query for finding entity version mapping.
+    */
+  private def getFilter(collectionName: String,
+                        entityName: String,
+                        oldVersion: Int): Bson = {
+    and(
+      and(
+        regex("collection", collectionName, "i"),
+        regex("entityName", entityName, "i")),
```

Comment:
is there a reason to have two levels of `and`?

Reply:
Good spot! I have several instances of these.

Reply:
This is actually not the only place if I remember correctly